storage/src/main/java/org/apache/kafka/storage/internals/log/LazyIndex.java

```diff
@@ -166,20 +166,25 @@ public File file() {
 
     @SuppressWarnings("unchecked")
     public T get() throws IOException {
-        if (indexWrapper instanceof IndexValue<?>)
-            return ((IndexValue<T>) indexWrapper).index;
-        else if (indexWrapper instanceof IndexFile) {
+        IndexWrapper wrapper = indexWrapper;
+        if (wrapper instanceof IndexValue<?>)
+            return ((IndexValue<T>) wrapper).index;
+        else {
             lock.lock();
             try {
-                IndexFile indexFile = (IndexFile) indexWrapper;
-                IndexValue<T> indexValue = new IndexValue<>(loadIndex(indexFile.file));
-                indexWrapper = indexValue;
-                return indexValue.index;
+                if (indexWrapper instanceof IndexValue<?>)
+                    return ((IndexValue<T>) indexWrapper).index;
```

Comment:
The bug is that we did not check inside the lock if `indexWrapper` was of type `IndexValue`.